Vagrantfile

```diff
@@ -6,16 +6,29 @@ load File.join(
 )
 
 Vagrant.configure(2) do |config|
-	config.vm.hostname = "blockextend"
+	config.vm.hostname = "block-scaffolding-wp"
 
 	# Wait to ensure all containers are up.
 	config.vm.provision "shell",
 		inline: "sleep 10",
 		run: "always"
 
+	# Install dev dependencies.
+	config.vm.provision "shell",
```

Comment:
This Vagrant wrapper serves mainly to isolate the Docker network used by the development environment and enable the `.local` URLs without having to change the hosts file. It is very similar to [Docker Machine](https://docs.docker.com/machine/overview/) but uses Vagrant for talking to VirtualBox which is more familiar to WP devs.

For consistency it would be great to stick to Docker containers for everything. This would also allow us to re-use them for things like GitHub actions, CI checks, etc. We could either bundle a custom Dockerfile that adds everything on top of the existing WordPress image or create our own Docker image repository that has a container with Composer and Node.js installed. For now, it might be easier to start with a bundled Dockerfile for building the all-encompassing container. Something like:

```Dockerfile
FROM wordpress:php7.3-apache

# Development tooling dependencies
RUN apt-get update; \
	apt-get install -y --no-install-recommends \
	bash less default-mysql-client git \
	nodejs npm curl

RUN curl -s https://getcomposer.org/installer | php \
	&& mv composer.phar /usr/local/bin/composer

# Setup xdebug.
RUN	pecl install xdebug; \
	docker-php-ext-enable xdebug;
```

Our task runners (npm & composer) could include helper scripts for running tests and building things inside the Docker containers either directly if the Docker engine is running on the host or through the Vagrant wrapper for those who prefer that. For example, here is a command to run phpunit inside the custom WordPress container with composer installed:

```
docker-compose exec wordpress composer --working-dir=wp-content/plugins/plugin-name test
```

which can be wrapped into a helper script:

```json
"scripts": {
    "docker-composer": "docker-compose exec -T wordpress composer --working-dir=wp-content/plugins/plugin-name",
}
```

so that running tests inside the Docker container is as simple as `composer docker-composer -- test`. We can go further by creating a generic wrapper for running any of the supported commands via:

```json
"scripts": {
    "docker": "docker-compose exec -T --workdir /var/www/html/wp-content/plugins/plugin-name wordpress",
}
```

so that we can do both npm and composer scripts:

```
composer docker -- composer test
composer docker -- npm install
```

This would work with both Composer and npm as the task runner so we just need to pick one. I suggest keeping the PHP-only tasks defined in `composer.json` while the rest of scripts are defined in `npm` since Composer scripts don't support interactive shell (TTY).

Finally, for the Vagrant wrapper we need to get more verbose because it's essentially two levels down:

```json
"scripts": {
    "vagrant": "vagrant ssh -- COMPOSE_FILE=/vagrant/docker-compose.yml docker-compose exec -T --workdir /var/www/html/wp-content/plugins/plugin-name wordpress"
}
```

so that all commands become `composer vagrant -- composer test`, etc.

## tldr:

Keep using Docker containers for encapsulating all development environment dependencies and use wrapper scripts for running things inside those containers:

```json
"scripts": {
    "docker": "docker-compose exec -T --workdir /var/www/html/wp-content/plugins/plugin-name wordpress",
    "vagrant": "vagrant ssh -- COMPOSE_FILE=/vagrant/docker-compose.yml docker-compose exec -T --workdir /var/www/html/wp-content/plugins/plugin-name wordpress"
}
```

so that the developer UX becomes:

```
composer docker -- composer test
composer vagrant -- composer test
```